File: 105. Construct Binary Tree from Preorder and Inorder Traversal/main.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func buildTree(preorder []int, inorder []int) *TreeNode {
	n := len(preorder)
	return traverse(preorder, inorder, 0, n-1, 0, n-1)
}

func traverse(preorder []int, inorder []int, preStart int, preEnd int, inStart int, inEnd int) *TreeNode {
	if preStart > preEnd {
		return nil
	}

	root := preorder[preStart]

	rootPos := -1
	for i, node := range inorder {
		if node == root {
			rootPos = i
			break
		}
	}

	leftSize := rootPos - inStart

	node := TreeNode{
		Val: root,
	}

	node.Left = traverse(preorder, inorder, preStart+1, preStart+leftSize, inStart, rootPos-1)
	node.Right = traverse(preorder, inorder, preStart+leftSize+1, preEnd, rootPos+1, inEnd)

	return &node
}

File: 105. Construct Binary Tree from Preorder and Inorder Traversal/main.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
from collections import deque
from typing import List, Optional


class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder:
            return None
        root = TreeNode(preorder[0])
        stack = [root]
        inPtr = 0

        for i in range(1, len(preorder)):
            preOrderVal = preorder[i]
            node = stack[-1]
            if node.val != inorder[inPtr]:
                node.left = TreeNode(preorder[i])
                stack.append(node.left)
            else:
                while stack and stack[-1].val == inorder[inPtr]:
                    node = stack.pop()
                    inPtr += 1
                node.right = TreeNode(preOrderVal)
                stack.append(node.right)
        
        return root

s = Solution()
s.buildTree([1,2,4,5,3,6,7], [4,2,5,1,6,3,7])